plugins/experimental/ja3_fingerprint/ja3_fingerprint.cc

```diff
@@ -0,0 +1,489 @@
+#include <unordered_map>
+#include <memory>
+#include <regex>
+
+#include "ts/ts.h"
+#include "ts/remap.h"
+
+#include <openssl/ssl.h>
+#include <openssl/md5.h>
+#include <openssl/opensslv.h>
+
+// Get 16bit big endian order and update pointer
+#define n2s(c, s) ((s = (((unsigned int)(c[0])) << 8) | (((unsigned int)(c[1])))), c += 2)
+
+const char *PLUGIN_NAME = "ja3_fingerprint";
+static TSTextLogObject pluginlog;
+static int ja3_idx    = -1;
+static int enable_raw = 0;
+static int enable_log = 0;
+
+// GREASE table as in ja3
+static const std::unordered_set<uint16_t> GREASE_table = {0x0a0a, 0x1a1a, 0x2a2a, 0x3a3a, 0x4a4a, 0x5a5a, 0x6a6a, 0x7a7a,
+                                                          0x8a8a, 0x9a9a, 0xaaaa, 0xbaba, 0xcaca, 0xdada, 0xeaea, 0xfafa};
+
+struct ja3_data {
+  std::string ja3_string;
+  char md5String[33];
+  char ip_addr[INET6_ADDRSTRLEN];
+};
+
+struct ja3_remap_info {
+  int raw        = false;
+  int log        = false;
+  TSCont handler = nullptr;
+
+  ~ja3_remap_info()
+  {
+    if (handler) {
+      TSContDestroy(handler);
+      handler = nullptr;
+    }
+  }
+};
+
+static int
+custom_get_ja3_prefixed(int unit, const unsigned char *&data, int len, std::string &result)
+{
+  int cnt, tmp;
+  bool first = true;
+  // Extract each entry and append to result string
+  for (cnt = 0; cnt < len; cnt += unit) {
+    if (unit == 1) {
+      tmp = *(data++);
+    } else {
+      n2s(data, tmp);
+    }
+
+    // Check for GREASE for 16-bit values, append only if non-GREASE
+    if (unit != 2 || GREASE_table.find(tmp) == GREASE_table.end()) {
+      if (!first) {
+        result += '-';
+      }
+      first = false;
+      result += std::to_string(tmp);
+    }
+  }
+  return 0;
+}
+
+char *
+getIP(sockaddr const *s_sockaddr, char res[INET6_ADDRSTRLEN])
+{
+  res[0] = '\0';
+
+  if (s_sockaddr == nullptr) {
+    return nullptr;
+  }
+
+  switch (s_sockaddr->sa_family) {
+  case AF_INET: {
+    const struct sockaddr_in *s_sockaddr_in = reinterpret_cast<const struct sockaddr_in *>(s_sockaddr);
+    inet_ntop(AF_INET, &s_sockaddr_in->sin_addr, res, INET_ADDRSTRLEN);
+  } break;
+  case AF_INET6: {
+    const struct sockaddr_in6 *s_sockaddr_in6 = reinterpret_cast<const struct sockaddr_in6 *>(s_sockaddr);
+    inet_ntop(AF_INET6, &s_sockaddr_in6->sin6_addr, res, INET6_ADDRSTRLEN);
+  } break;
+  default:
+    return nullptr;
+  }
+
+  return res[0] ? res : nullptr;
+}
+
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+// Parsing clientHello to get ja3 string
+// No error checking or handling because this should be called after openSSL has done all checks and
+// returned successfully
+static std::string
+custom_get_ja3(SSL *s)
+{
+  TSDebug(PLUGIN_NAME, "Entering custom_get_ja3()...");
+  std::string ja3;
+  const unsigned char *p, *d;
+  int i, j, len;
+
+  // ClientHello buf and len
+  d = p  = (unsigned char *)s->init_msg;
+  long n = s->init_num;
+
+  // Get version
+  int version = (((int)p[0]) << 8) | (int)p[1];
```

Comment:
Isn't this the 'n2s' function? Alternatively, which is this one signed? That would seem to be problematic if `p[0]` has the high bit set.

Reply:
These 2 bytes only have 4 valid combinations:
SSL3_VERSION x’0300′
TLS1_VERSION x’0301′
TLS2_VERSION x’0302′
TLS3_VERSION x’0303′